File: interview_cake/arrays_and_strings/reverse_words.js
function reverseWords(message) {

  // Decode the message by reversing the words
 message = reverseChars(message, 0, message.length - 1);
 
 word_start = 0;
 for ( let i = 0; i <= message.length; i++ ) {
   if ((message[i] == ' ') || (i == message.length)) {
     reverseChars(message, word_start, i - 1);
     word_start = i + 1;
   }
 }
}

function reverseChars(string, start_index, end_index) {
  while (end_index > start_index) {
    placeholder = string[end_index];
    string[end_index] = string[start_index];
    string[start_index] = placeholder;
    start_index += 1;
    end_index -= 1;
  }
  
  return string;
}
















// Tests

let desc = 'one word';
let input = 'vault'.split('');
reverseWords(input);
let actual = input.join('');
let expected = 'vault';
assertEqual(actual, expected, desc);

desc = 'two words';
input = 'thief cake'.split('');
reverseWords(input);
actual = input.join('');
expected = 'cake thief';
assertEqual(actual, expected, desc);

desc = 'three words';
input = 'one another get'.split('');
reverseWords(input);
actual = input.join('');
expected = 'get another one';
assertEqual(actual, expected, desc);

desc = 'multiple words same length';
input = 'rat the ate cat the'.split('');
reverseWords(input);
actual = input.join('');
expected = 'the cat ate the rat';
assertEqual(actual, expected, desc);

desc = 'multiple words different lengths';
input = 'yummy is cake bundt chocolate'.split('');
reverseWords(input);
actual = input.join('');
expected = 'chocolate bundt cake is yummy';
assertEqual(actual, expected, desc);

desc = 'empty string';
input = ''.split('');
reverseWords(input);
actual = input.join('');
expected = '';
assertEqual(actual, expected, desc);

function assertEqual(a, b, desc) {
  if (a === b) {
    console.log(`${desc} ... PASS`);
  } else {
    console.log(`${desc} ... FAIL: ${a} != ${b}`);
  }
}

File: interview_cake/reverse_words.rb
def reverse_words!(message)

  # Decode the message by reversing the words.
  
  reverse_chars(message, 0, message.length - 1)
  
  word_start = 0 
  (0..message.length).each do |i|
    if message[i] == ' ' || i == message.length
      reverse_chars(message, word_start, i - 1)
      word_start = i + 1
    end
  end
end

def reverse_chars(string, start_index, end_index)
  while end_index > start_index 
    placeholder = string[start_index]
    string[start_index] = string[end_index]
    string[end_index] = placeholder
    start_index += 1
    end_index -= 1
  end
end


















# Tests

def run_tests
  desc = 'one word'
  message = 'vault'
  reverse_words!(message)
  expected = 'vault'
  assert_equal(message, expected, desc)

  desc = 'two words'
  message = 'thief cake'
  reverse_words!(message)
  expected = 'cake thief'
  assert_equal(message, expected, desc)

  desc = 'three words'
  message = 'one another get'
  reverse_words!(message)
  expected = 'get another one'
  assert_equal(message, expected, desc)

  desc = 'multiple words same length'
  message = 'rat the ate cat the'
  reverse_words!(message)
  expected = 'the cat ate the rat'
  assert_equal(message, expected, desc)

  desc = 'multiple words different lengths'
  message = 'yummy is cake bundt chocolate'
  reverse_words!(message)
  expected = 'chocolate bundt cake is yummy'
  assert_equal(message, expected, desc)

  desc = 'empty string'
  message = ''
  reverse_words!(message)
  expected = ''
  assert_equal(message, expected, desc)
end

def assert_equal(a, b, desc)
  puts "#{desc} ... #{a == b ? 'PASS' : "FAIL: #{a.inspect} != #{b.inspect}"}"
end

run_tests
